tutorials/Oracles/Oracles.ipynb

```diff
@@ -0,0 +1,1015 @@
+    "operation Phase_Oracle_Demo() : Unit {\n",
+    "    let divider = \"--------------------------------------------------------------------------------------------------\";\n",
+    "    \n",
+    "    // allocate the |000⟩ state\n",
+    "    using(q = Qubit[3]) {\n",
+    "        // Prepare an equal superposition of all basis states\n",
+    "        ApplyToEachA(H, q);\n",
+    "        \n",
+    "        // Print out that we currently have a superposition of all the basis states\n",
+    "        // Notice the phases on each of the components\n",
+    "        Message(\"State |β⟩ (equal superposition of all basis states):\");\n",
+    "        DumpMachine();\n",
+    "        Message(divider);\n",
+    "        \n",
+    "        // Apply the oracle\n",
+    "        Alternating_Phase_Oracle(q);\n",
+    "        \n",
+    "        \n",
+    "        // Print out the resulting state of AND |β⟩\n",
+    "        // Notice that the phase of the state is still positive\n",
+    "        Message(\"State after applying the alternating phase oracle to |β⟩:\");\n",
+    "        DumpMachine();\n",
+    "        Message(divider);\n",
+    "        \n",
+    "        // Reset our state q back to all zeros for deallocation\n",
+    "        ResetAll(q);\n",
+    "    }\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "%config dump.basisStateLabelingConvention=\"Bitstring\"\n",
+    "%simulate Phase_Oracle_Demo"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "> Notice that the input state above is an equal superposition and after applying the oracle we have the same equal superposition except that the states $|010\\rangle$ and $|101\\rangle$ had their phase flipped!  Recall that these two states are exactly the two states for which $f(x) = 1$ for the alternating bit pattern phase oracle $U_{alt,phase}$, thus they are exactly the two states we expect to experience a phase flip!"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now you will implement the same classical oracle that you implemented in <span style=\"color:blue\">Exercise 1</span> as a quantum phase oracle.\n",
+    "\n",
+    "#### <span style=\"color:blue\">Exercise 2</span>: Implement a phase quantum oracle\n",
+    "Implement the following quantum oracle $U_{7,phase}$:\n",
+    "* $U_{7,phase} |x\\rangle = (-1)^{f(x)}|x\\rangle$\n",
+    "* $f(x) = 1$ if $x = 7$\n",
+    "* $f(x) = 0$ otherwise\n",
+    "* $|x\\rangle$ is a state composted of $3$ qubits\n",
+    "\n",
+    "<br/>\n",
+    "<details>\n",
+    "  <summary><b>Need a hint? Click here</b></summary>\n",
+    "  What internal state can you prepare such that $X|\\psi\\rangle = -|\\psi\\rangle$ or $Z|\\psi\\rangle = -|\\psi\\rangle$?  Finally use this internal state to flip the sign of the input state $|x\\rangle$.\n",
```

Comment:
I tried to figure out a solution based on this hint and I'm really not sure whether the learner would be able to, since applying just an X or a Z gate won't help them... I would try something like "You need to find a gate that will flip the sign of the |111> basis state but not the other states. Which single-qubit gate flips the sign of the |1> basis state but not the |0>? How to modify it to solve this problem?"

Reply:
Yes, my hint was suggesting to allocate auxiliary quits to assist with the computation but the exercise was updated to require that the learner does not allocate more cubits and I overlooked changing the hint.  The hint you have provided is great as it promotes the idea of thinking about control and the Z gate.